### PR TITLE
GGUF - QK4_0 quantization for LLaMA

### PR DESCRIPTION
**Pull request by kunwar31**
output looks similar with manual testing, but lm_eval eventually OOMs after running for some time. 

tiny llama without quantization
```
time GPU=1 python examples/llama.py --temperature 0 --gen "tiny" --size "1B-Chat" --count 100 --model "./weights/LLaMA/1B-Chat/model.safetensors" --prompt="Hello."
using GPU backend
using LLaMA-tiny-1B-Chat model
ram used:  4.40 GB, freqs_cis                                         : 100%|█████████████████████████████████████████████████████████████████████| 202/202 [00:00<00:00, 243.62it/s]
loaded weights in 833.33 ms, 4.40 GB loaded at 5.28 GB/s
Hello. I am trying to create a simple Hello World program in C. I have written the following code:
HelloWorld.c
#include <stdio.h>

int main() {
    printf("Hello World!\n");
    return 0;
}

When I compile it with gcc -o HelloWorld HelloWorld.c -o HelloWorld, I get the following output:
$ gcc -o HelloWorld HelloWorld.c
$ ./HelloWorld

real	0m4.888s
user	0m4.592s
sys	0m1.667s
```
tiny llama with 4bit GGUF QK4_0 quantization
```
time GPU=1 python examples/llama.py --temperature 0 --gen "tiny" --size "1B-Chat" --count 100 --model "./weights/LLaMA/1B-Chat/model.safetensors" --prompt="Hello." --quantize --use_4bit
using GPU backend
using LLaMA-tiny-1B-Chat model
ram used:  0.84 GB, freqs_cis                                         : 100%|███████████████████████████████████████████████████████████████████| 357/357 [00:00<00:00, 15614.16it/s]
loaded weights in 25.85 ms, 0.00 GB loaded at 0.00 GB/s
Hello.
I am trying to create a simple Hello World program in C.
I have tried the following code but it does not work:
#include <stdio.h>

int main() {
    printf("Hello World!\n");
    return 0;
}

Can you please help me?
Thank you.

A: #include <stdio.h>

int main() {
    printf("Hello World!\n");
    return 
real	0m10.370s
user	0m10.405s
sys	0m1.607s
```

### CODE REVIEW

**Comment by kunwar31** [comment]
on RTX 3060, I get 35 tok/s without quantize and 14 tok/s with 4bit quantize

**Comment by geohot** [comment]
Has to load from the gguf file for the bounty

**Comment by aleir97** [comment]
We are now able to load both tokenizer and weights from a 4 bit .gguf model. Only linear layers are dequantized on the fly and we are still working on performance to get a better token rate at a 8Gb RAM on a M1 chip.
``` console
JIT=1 python3 examples/llama.py --temperature 0 --gen "2" --count 100 --model "./weights/llama-2-7b.Q4_0.gguf" --prompt="Chicken is  "
```

**Comment by torphix** [comment]
Is this the bounty for loading llama2-7b-gguf file?
I wrote the code for it yesterday as spreadsheet said it was unclaimed, is this still live or should i make a pull request?

**Comment by g1y5x3** [comment]
There's a lock tag on this PR already. The spreadsheet can be a bit behind.

**Comment by torphix** [comment]
Okay, ill be sure to check PRs first in the future. One thing I found that
might be helpful to speed up inference is if you cache some of quantized
weights eg: 1/8 of the layers

**Comment by aleir97** [comment]
@torphix thanks for the info! The main issue right now is RAM usage so.. is it worth it to cache more stuff?

**Comment by aleir97** [comment]
Working through this I found that the KV cache for large context lenghts is forcing the usage of swap memory at 8GB of RAM  so the performance drops a lot. Nonetheless at my M1 MacBook Air I'm getting around 4 tok/s BEAM=4, JIT=1 and TC=1. Waiting for @kunwar31 to merge some minor fixes.

**Comment by torphix** [comment]
@aleir97  Nice work, if KV cache is causing too much memory consumption consider experimenting with multi query attention (MQA) which can reduce memory consumption by ~20x or grouped query attention which alters the memory / accuracy trade off 
It shouldn't take too much alteration to the KV cache code as it is simpler version.
MQA: https://arxiv.org/abs/1911.02150
GQA: https://arxiv.org/abs/2305.13245

**Comment by aleir97** [comment]
Good news!! Doing the correct math we are getting almost 10 tok/s!! Last time I checked this is better than MLX llama.py example and matches llama.cpp
Captura de pantalla 2024-02-01 a las 15 35 56

**Comment by kunwar31** [comment]
@aleir97 you can have the bounty, I haven't worked much on this

**Comment by geohot** [comment]
Let me know when this is ready for review.

File: setup.py [hunk 1]
```diff
@@ -52,6 +52,8 @@
             "librosa",
             "networkx",
             "hypothesis",
+            "protobuf==3.20.3",
+            "gguf",
```
Comment (kunwar31) [inline setup.py:56]: @aleir97 does it need to be in setup?

**Comment by aleir97** [comment]
@geohot Just updated everything to master and cleaned the PR. To run the example: 
```console
MAX_CONTEXT=512 BEAM=4 JIT=1 python3 tinygrad/examples/llama.py --size=7B --gen=2 --temperature=0 --count=30 --prompt="Chicken is"  --model=tinygrad/weights/llama-2-7b.Q4_0.gguf 
```

**Comment by geohot** [comment]
Have you run it through some evaluation to confirm the output is correct and matches GGML?

File: examples/llama.py [hunk 2]
```diff
@@ -13,7 +13,9 @@
 from tinygrad.tensor import Tensor
 from tinygrad.nn.state import safe_load, torch_load, load_state_dict, get_parameters
 from extra.models.llama import Transformer, convert_from_huggingface
-from sentencepiece import SentencePieceProcessor
+from sentencepiece import SentencePieceProcessor, sentencepiece_model_pb2
+import gguf
```
Comment (geohot) [inline examples/llama.py:17]: I don't want to use this (certainly not required to run llama.py). Can we open this file format in state?

**Comment by geohot** [comment]
No updates in weeks, going to close. This needs testing to confirm it matches GGML
